src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java

```diff
@@ -0,0 +1,132 @@
+    }
+
+    private static class RemoveToStringFromArraysVisitor extends JavaVisitor<ExecutionContext> {
+        @Override
+        public J visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
+            if (OBJECT_TOSTRING_MATCHER.matches(mi) || OBJECTS_TOSTRING_MATCHER.matches(mi)) {
+                Expression select = mi.getSelect();
+
+                if (select != null) {
+                    if (!(select.getType() instanceof JavaType.Array)) {
+                        return mi;
+                    }
+
+                    maybeAddImport("java.util.Arrays");
+                    return JavaTemplate.builder("Arrays.toString(#{anyArray(java.lang.Object)})")
+                            .imports("java.util.Arrays")
+                            .build()
+                            .apply(getCursor(), mi.getCoordinates().replace(), select);
+                }
+            } else if (METHOD_MATCHERS.stream().anyMatch(matcher -> matcher.matches(mi))) {
+                // deals with edge cases where .toString() is called implicitly
+                List<Expression> arguments = mi.getArguments();
+                for (Expression arg : arguments) {
+                    if (arg.getType() instanceof JavaType.Array) {
+                        Cursor c = getCursor();
+                        c.putMessage("METHOD_KEY", mi);
+                        break;
+                    }
+                }
+            }
+
+            return super.visitMethodInvocation(mi, ctx);
+        }
+
+        @Override
+        public Expression visitExpression(Expression exp, ExecutionContext ctx) {
+            Expression e = (Expression) super.visitExpression(exp, ctx);
+            if (e.getType() instanceof JavaType.Array) {
+                Cursor c = getCursor().dropParentWhile(is -> is instanceof J.Parentheses || !(is instanceof Tree));
+                if (c.getMessage("METHOD_KEY") != null || c.getMessage("BINARY_FOUND") != null) {
+                    maybeAddImport("java.util.Arrays");
+                    return JavaTemplate.builder("Arrays.toString(#{anyArray(java.lang.Object)})")
+                            .imports("java.util.Arrays")
+                            .build()
+                            .apply(getCursor(), e.getCoordinates().replace(), e);
+                }
+            }
+
+            return e;
+        }
+
+        @Override
+        public J.Binary visitBinary(J.Binary binary, ExecutionContext ctx) {
+            Expression left = binary.getLeft();
+            Expression right = binary.getRight();
+
+            Cursor c = getCursor().dropParentWhile(is -> is instanceof J.Parentheses || !(is instanceof Tree));
+            if (c.getValue() instanceof J.MethodInvocation) {
+                if (PRINT_MATCHER.matches((J.MethodInvocation) c.getValue())) {
+                    if (left.getType() instanceof JavaType.Array || right.getType() instanceof JavaType.Array) {
```

Comment:
There is no need to constrain this to method invocations. So you can drop the two outer `if`s here and on this innermost `if` I would first also check the condition `binary.getOperator() == J.Binary.Type.Addition`.